### PR TITLE
Defined Commands and environment variables

### PR DESCRIPTION
**Pull request by AlbertDeFusco**
Here's the plan to address #2 

TODO:

- [x] tests (specifically, how can we test `os.execvpe`?)
- [x] variables in commands
- [x] #50 
- [x] `conda project activate` to launch a new shell with the env activated
- [x] ensure that run and activate work both in cmd.exe and powershell
- [ ] docs

```yaml
❯ cat conda-project.yml
name: cmds
environments:
  default:
    - environment.yml
  aux:
    - environment.yml
    - extras.yml

variables:
  FOO: bar

commands:
  versions: python vars.py
  others: 
    cmd: tranquilizer --version
    environment: aux
```

commands are defined to run against the default (first) environment spec unless that is manually specified. 

```python
❯ cat vars.py
import os
import sys

print(sys.version)
print(os.getenv('FOO'))
print(os.getenv('BAR'))
```

running the default command, called `versions` executes the `var.py` file shown above.

```
❯ conda project run
The lockfile /Users/adefusco/Development/conda-incubator/conda-project/examples/cmds/default.conda-lock.yml is out-of-date, re-locking...
Locking dependencies for default: done
3.9.13 (main, Aug 25 2022, 18:24:45) 
[Clang 12.0.0 ]
bar
None
```

If a variable is defined but no default value is given it will error

```yaml
❯ cat conda-project.yml
name: cmds
environments:
  default:
    - environment.yml
  aux:
    - environment.yml
    - extras.yml

variables:
  FOO: bar
  BAR:

commands:
  versions: python vars.py
  others: 
    cmd: tranquilizer --version
    environment: aux
```

```
❯ conda project run
CondaProjectError: The following variables do not have a default value and were not set when executing 'conda project run':
BAR
```

The value of a defined variable can be provided or overridden by the calling process

```
❯ FOO=not-default BAR=foo conda project run
3.9.13 (main, Oct 13 2022, 16:12:19) 
[Clang 12.0.0 ]
not-default
foo
```

### CODE REVIEW

**Comment by AlbertDeFusco** [comment]
This PR is in draft state until schema and behavior finalized and tests written.

**Comment by AlbertDeFusco** [comment]
One thing to consider here is multiline commands. Right now only single line commands will work. To me multiline commands begin implication of shell support and we would then end up in the anaconda-project scenario of separation of unix and windows commands. We would then have to consider replicating or vendoring `conda run`, which creates a temporary file with the commands to execute

https://github.com/conda/conda/blob/49b1bad3d70a5d2f2bc5be96b3cc62e475b92968/conda/utils.py#L346-L462

and then using subprocess to execute the temporary script

https://github.com/conda/conda/blob/49b1bad3d70a5d2f2bc5be96b3cc62e475b92968/conda/cli/main_run.py#L28-L35

We're already calling this whole procedure from conda-project. We _might_ be able to transform a multi-line command into something that will work here.

https://github.com/conda-incubator/conda-project/blob/1ba9cc3df6941021d59d020ce18435ebe8fcb651/conda_project/models.py#L348-L353

**Comment by mattkram** [comment]
I really like the design. And I agree that it would be great if possible to allow multi-line commands that are cross-platform. It would be good to understand the constraints learned from `anaconda-project` and figure out if there's a way (and what is the use case) to do it as cross-platform as possible.

Another feature, which we can defer until later, might be to allow listing of predecessor commands. It would sort of function as a minimal `Makefile`, or like `nox` session. E.g. you could have `conda project run test`, which maybe requires `lint`, `fmt`, etc. That's more from the dev perspective, but it could potentially be useful in other contexts. Again, that could probably be a follow-in feature to the core functionality.

**Comment by mattkram** [comment]
> We've heard a need for dependencies. Something like this makes sense to me. 
> 
> ```yaml
> commands:
>   last:
>     cmd: <>
>     requires:
>       - first
>       - middle
>   first: <>
>   middle: <>
> ```

Totally agree. And probably just specify that ordering would be lexically sorted if precedence resulted in a tie.

I've got a version of this in another early project [here](https://lembas.readthedocs.io/en/latest/user_guide.html#case-steps) if we want to steal some implementation.

**Comment by AlbertDeFusco** [comment]
This PR is a great opportunity to provide environment activation support as well. Does anyone here have strong opinions around the follow scenarios? Or are there others we can consider?

* open a subshell with env activated and any variables from the project file set; you would run `exit` to get back to previous env rather than `deactivate`; this is more like `poetry shell` than `conda activate`
* make `conda activate ./envs/default` work natively while also setting any variables from the project file
* have `conda project activate` work just like `conda activate <prefix>` rather than opening a new subshell while setting environment variables defined in the project yaml

**Comment by mattkram** [comment]
> This PR is a great opportunity to provide environment activation support as well. Does anyone here have strong opinions around the follow scenarios? Or are there others we can consider?
> 
> * open a subshell with env activated and any variables from the project file set; you would run `exit` to get back to previous env rather than `deactivate`; this is more like `poetry shell` than `conda activate`
> * make `conda activate ./envs/default` work natively while also setting any variables from the project file
> * have `conda project activate` work just like `conda activate <prefix>` rather than opening a new subshell while setting environment variables defined in the project yaml

I think all of these would be valid use cases. However options 1&3 are similar and I think I'd prefer 3.

We would want to be able to provide the environment as an option or argument, i.e. `conda project activate dev` or `conda project activate --env dev`.

**Comment by jlstevens** [comment]
I don't like option 1 much but I think options 2 and 3 can both exist together. I'm not sure I would expect `conda activate` to respect projects, but it is a nice bonus if it does do the right thing.

Therefore I would probably use the following:

```
conda project activate # I would assume this activates the default env when nothing is specified
# OR
conda project activate <named env or prefix> 
# e.g. `conda project activate test` or `conda project activate ./envs/test` should both work
```

Just my 2c.

**Comment by AlbertDeFusco** [comment]
Thanks for that. Certainly I agree that the default (first) env is activated and `conda [project] activate [env-name]` allows other defined envs to activate. The trick is actually making it work since `conda activate` is actually a shell command, not running in python. I'll keep digging and report back

**Comment by AlbertDeFusco** [comment]
I just found this either as down-the-road inspiration or perhaps something we could integrate or utilize

https://taskfile.dev/

**Comment by mattkram** [comment]
> I just found this either as down-the-road inspiration or perhaps something we could integrate or utilize
> 
> https://taskfile.dev/

That's pretty neat. I've used invoke in the past, which is python-based.

https://www.pyinvoke.org/

**Comment by AlbertDeFusco** [comment]
TL;DR I've implemented the subshell approach for now. The subshell approach also helps set project-defined variables. I'm looking at being able to get them in the `<prefix>/etc/conda/[de]activate.d` collection so that `conda activave ./env/<name>` will work.

## conda project activate command
So in my research I've learned that `conda` is actually a [shell command](https://github.com/conda/conda/blob/main/conda/shell/etc/profile.d/conda.sh). This means that `conda activate` does the following (and something similar on windows) directly in the shell and not in the Python runtime

1. executes `VARS_TO_SET=$(/path/to/conda/executable shell.posix activate ${env_name_or_prefix})` to get the env variables
    * Try this yourself to see what `conda activate` actually does
3. `eval "$VARS_TO_SET"` to officially activate the environment

The upshot is that it's not possible to subprocess this procedure in Python to then interact with the parent shell. We could replicate the conda's shell procedures. It might required a `conda project init`, which seems overkill to me right now.

Perhaps in the future with Conda Plugins we'll be able to extend the shell commands to enable special handling for `conda project activate [env-name]`.

**Comment by jlstevens** [comment]
This looks great! 

I'll see if I can checkout this PR and give it a try. Before I do, I have two quick questions I didn't manage to answer for myself after a quick glance at the diff:

1. What is the plan around `supports_http_options` (or equivalent)? 
2. On a related note, is the jinja2 templating approach I implemented for anaconda-project still around? This approach can help support http options...
